Make a static pinning decision based on the last level with data (not bottommost level) (#626)

udi-speedb opened the pull request.

mrambacher commented on db/column_family.cc:
```diff
@@ -1658,6 +1661,24 @@ void ColumnFamilyData::RecoverEpochNumbers() {
+  return current_->storage_info();
+}
+
+int ColumnFamilyData::IsLastLevelWithData(int level) const {
```
bool?

udi-speedb replied:
Done

mrambacher commented on db/column_family.cc:
```diff
@@ -1658,6 +1661,24 @@ void ColumnFamilyData::RecoverEpochNumbers() {
+  int last_level_with_data = vstorage->num_non_empty_levels() - 1;
+
+  auto is_last_level_with_data = false;
+  if ((level > 0) and (level == last_level_with_data)) {
```
"&&" rather than "and" ?

udi-speedb replied:
Done

mrambacher commented on db/table_cache.cc:
```diff
@@ -68,15 +68,17 @@ TableCache::TableCache(const ImmutableOptions& ioptions,
                        const std::shared_ptr<IOTracer>& io_tracer,
-                       const std::string& db_session_id)
+                       const std::string& db_session_id,
+                       IsLastLevelWithDataFunc is_last_leve_with_data_func)
```
"is_last_level_with_data_func" (not leve_)

udi-speedb replied:
Done

mrambacher commented on include/rocksdb/table_pinning_policy.h:
```diff
@@ -21,27 +21,32 @@ struct ConfigOptions;
 
-  TablePinningOptions(int _level, bool _is_bottom, size_t _file_size,
-                      size_t _max_file_size_for_l0_meta_pin)
+  TablePinningOptions(int _level, bool _is_last_level_with_data,
+                      size_t _file_size, size_t _max_file_size_for_l0_meta_pin)
```
Is there a reason not to have both "last level with data" and "bottom"?  I know the current use case does not seem to care about the distinction but I am wondering if it could be important in the future (or if this use case proves to be incomplete).

udi-speedb replied:
Hilik's intention is clearly to decide based on the last level with data. 
If we ever do find a use case for the bottom flag as well, we will add it. I really see no need to have it just in case.

udi-speedb commented:
@mrambacher I have pushed a commit addressing the issues you have raised.
Are you done with the review?
I need this PR for #660 so, I really need this PR merges soon.Than

mrambacher commented on db/db_test.cc:
```diff
@@ -7412,6 +7414,130 @@ TEST_F(DBTest, ShuttingDownNotBlockStalledWrites) {
+  }
+
+  void UnPinData(std::unique_ptr<PinnedEntry>&& pinned) override {
+    ASSERT_GT(total_num_pinned_, 0U);
+    --total_num_pinned_;
+
+    ASSERT_GE(usage_, pinned->size);
+    usage_ -= pinned->size;
+
+    if (pinned->is_last_level_with_data) {
+      ASSERT_GT(num_pinned_last_level_with_data_, 0U);
```
These should probably be EXPECT_xx rather than ASSERT_xx ?

udi-speedb replied:
Why shouldn't it be ASSERT_xx?

mrambacher replied:
> Why shouldn't it be ASSERT_xx?

In my experience, ASSERT should be when the top-level method is failing whereas EXPECT should be when the failure is a long way from the code.  

I would suggest you run an experiment where you hit one of these inner ASSERT statements and observe how the test fails.  I often find those deep ASSERT messages hard to trace

udi-speedb replied:
I see, however, the guideline I am following when choosing EXPECT_* rather than ASSERT_* is the following:
`"The rule of thumb is to use EXPECT_* when you want the test to continue to reveal more errors after the assertion failure, and use ASSERT_* when continuing after failure doesn’t make sense"` ([http://google.github.io/googletest/primer.html#same-data-multiple-tests](url))

mrambacher commented on db/table_cache.h:
```diff
@@ -49,12 +50,16 @@ class HistogramImpl;
 // ioptions.row_cache
 class TableCache {
+ public:
+  using IsLastLevelWithDataFunc = std::function<int(int level)>;
```
The function should return a bool

udi-speedb replied:
Done

mrambacher reviewed (approved):
Approved with minor comments/suggestions

udi-speedb commented:
@mrambacher Thanks for the review.
I have updated per your last comment, squashed, and forced pushed.

Yuval-Ariel commented:
@udi-speedb , the CI seems to have failed but the force push removed that commit from this PRs history.. please refrain from force pushing after approval or rerun the CI if you do. 
https://github.com/speedb-io/speedb/actions/runs/6302889926/job/17110975186 
```
ninja: job failed: ccache /usr/bin/clang++ -DBZIP2 -DGFLAGS=1 -DGFLAGS_IS_A_DLL=0 -DHAVE_PCLMUL -DHAVE_SSE42 -DLZ4 -DOS_LINUX -DROCKSDB_FALLOCATE_PRESENT -DROCKSDB_LIB_IO_POSIX -DROCKSDB_MALLOC_USABLE_SIZE -DROCKSDB_NO_DYNAMIC_EXTENSION -DROCKSDB_PLATFORM_POSIX -DROCKSDB_PORTABLE -DROCKSDB_RANGESYNC_PRESENT -DROCKSDB_SCHED_GETCPU_PRESENT -DSNAPPY -DZLIB -DZSTD -I../ -I../include -isystem ../third-party/gtest-1.8.1/fused-src -W -Wextra -Wall -pthread -Wsign-compare -Wshadow -Wno-unused-parameter -Wno-unused-variable -Woverloaded-virtual -Wnon-virtual-dtor -Wno-missing-field-initializers -Wno-strict-aliasing -Wno-invalid-offsetof -Werror -g -DROCKSDB_USE_RTTI -std=gnu++17 -MD -MT CMakeFiles/db_test.dir/db/db_test.cc.o -MF CMakeFiles/db_test.dir/db/db_test.cc.o.d -o CMakeFiles/db_test.dir/db/db_test.cc.o -c ../db/db_test.cc
../db/db_test.cc:7462:15: error: 'ToString' overrides a member function but is not marked 'override' [-Werror,-Winconsistent-missing-override]
  std::string ToString() const {
              ^
../include/rocksdb/table_pinning_policy.h:94:23: note: overridden virtual function is here
  virtual std::string ToString() const = 0;
                      ^
1 error generated.
```

udi-speedb commented:
@Yuval-Ariel 

1. Can't refrain from forced push. PR was approved and I had to squash my commits. We need to decide what to do about forced push.
2. How come tests have passed before the forced push? Unless I am mistaken, the changes introduced by the forced push are not those causing the CI to fail now.

udi-speedb commented:
@Yuval-Ariel As I have already noted. If a forced push introduces code changes (not squash / commit message changes only), the CI should automatically restart and merge should be blocked as long as it doesn't pass the CI

udi-speedb commented:
@Yuval-Ariel - Manual run of the CI passed ([https://github.com/speedb-io/speedb/actions/runs/6312583680/job/17138928679](url))

Please merge

Yuval-Ariel commented:
thats actually the main branch (fc2bee8a1c51d66b85f3f8818cd5df8aad24f3f2).. i've run it on 6784aadcf1b855d8ba92f3c512d69a6fcc45ed50 and it passed.
in any case, shouldnt this go through performance before it can be merged?

udi-speedb commented:
> thats actually the main branch ([fc2bee8](https://github.com/speedb-io/speedb/commit/fc2bee8a1c51d66b85f3f8818cd5df8aad24f3f2)).. i've run it on [6784aad](https://github.com/speedb-io/speedb/commit/6784aadcf1b855d8ba92f3c512d69a6fcc45ed50) and it passed. in any case, shouldnt this go through performance before it can be merged?

I see. Well, It's always good to know that our main branch passes the CI.
I will make sure I know how to run the CI on the desired branch next time :-)

As for performance, you are correct. Thanks

udi-speedb commented:
@erez-speedb Please run performace tests on this branch. Thanks

udi-speedb commented:
@erez-speedb - Please update on the results of the performance testing and if we may merge this pr or do something else. Thanks
